sdk/storage/src/core/authorization_policy.rs

```diff
@@ -0,0 +1,256 @@
+use crate::clients::{ServiceType, StorageCredentials};
+use crate::headers::CONTENT_MD5;
+use azure_core::{headers::MS_DATE, Context, Policy, PolicyResult, Request, Response};
+use http::header::AUTHORIZATION;
+use http::HeaderValue;
+use http::{header::*, method::Method};
+use ring::hmac;
+use std::sync::Arc;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AuthorizationPolicy {
+    storage_credentials: StorageCredentials,
+}
+
+impl AuthorizationPolicy {
+    pub(crate) fn new(storage_credentials: StorageCredentials) -> Self {
+        Self {
+            storage_credentials,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Policy for AuthorizationPolicy {
+    async fn send(
+        &self,
+        ctx: &Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult<Response> {
+        if next.is_empty() {
+            return Err(Box::new(azure_core::PipelineError::InvalidTailPolicy(
+                "Authorization policies cannot be the last policy of a pipeline".to_owned(),
+            )));
+        }
+
+        match &self.storage_credentials {
+            StorageCredentials::Key(account, key) => {
+                let url = url::Url::parse(&request.uri().to_string()).unwrap();
```

Comment:
This is a bit awkward. I wonder if there's a better way to do this.

Reply:
this is actually a point I stumbled across when developing. Right now the SAS authorization does not work via the authorization policy, since I could not figure out a straight forward way to mutate the url within the policy. I guess the same applies here. So any ideas to mitigate this would be great.